opencog/atomspace/AtomTable.h

```diff
@@ -232,12 +242,26 @@ class AtomTable
                         bool subclass = false) const
     {
         std::lock_guard<std::recursive_mutex> lck(_mtx);
+
+        // We're going to use this one handle so we don't construct/destruct
+        // googles of temporaries in the process of doing the copies.
+        Handle temp_handle;
         std::for_each(typeIndex.begin(type, subclass),
                       typeIndex.end(),
-             [&](Handle h)->void {
-                  if (not isDefined(h)) return;
-                  (func)(h);
-             });
+            [&](UUID uuid)->void {
+                if (UNDEFINED_UUID == uuid) return;
+                
+                // Since the function could and probably will resolve the
+                // AtomPtr, we need to reset the handle since we are going to
+                // bypass the normal shared_ptr semantics here to avoid
+                // a copy per handle which is much more expensive than
+                // resetting and setting a new UUID. So we call the safe
+                // version which takes a bit longer.
+                temp_handle.set_uuid(uuid);
```

Comment:
The gcc and the clang optimizers easily optimize away calls to constructors when the constructors are relatively simple. Thus, a typical call to a consructor on stack becomes just an increment of the stack pointer by the object size, and possibly some initialization of parts of it.